File: src/main/java/me/zeroeightsix/kami/command/commands/ToggleCommand.java
package me.zeroeightsix.kami.command.commands;

import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import com.mojang.brigadier.builder.RequiredArgumentBuilder;
import me.zeroeightsix.kami.command.Command;
import me.zeroeightsix.kami.command.KamiCommandSource;
import me.zeroeightsix.kami.command.ModuleArgumentType;
import me.zeroeightsix.kami.module.Module;
import net.minecraft.server.command.CommandSource;

import static net.minecraft.util.Formatting.*;
import static me.zeroeightsix.kami.util.Texts.*;

/**
 * Created by 086 on 17/11/2017.
 */
public class ToggleCommand extends Command {

    public static boolean modifiersEnabled = true; // TODO: move this somewhere appropriate

    @Override
    public void register(CommandDispatcher<CommandSource> dispatcher) {
        dispatcher.register(
            LiteralArgumentBuilder.<CommandSource>literal("toggle")
                .then(
                    RequiredArgumentBuilder.<CommandSource, Module>argument("module", ModuleArgumentType.module())
                        .executes(context -> {
                            Module m = context.getArgument("module", Module.class);
                            m.toggle();
                            ((KamiCommandSource) context.getSource()).sendFeedback(
                                    f(GOLD, append(
                                            lit("Toggled module "),
                                            flit(YELLOW, m.getName()),
                                            lit(", now "),
                                            flit(m.isEnabled() ? GREEN : RED, m.isEnabled() ? "ON" : "OFF")
                                    ))
                            );
                            return 0;
                        })
                )
        );
    }
}

File: src/main/kotlin/me/zeroeightsix/kami/command/commands/ToggleCommand.kt
package me.zeroeightsix.kami.command.commands

import com.mojang.brigadier.CommandDispatcher
import com.mojang.brigadier.builder.LiteralArgumentBuilder
import com.mojang.brigadier.builder.RequiredArgumentBuilder
import com.mojang.brigadier.context.CommandContext
import me.zeroeightsix.kami.command.Command
import me.zeroeightsix.kami.command.KamiCommandSource
import me.zeroeightsix.kami.command.ModuleArgumentType
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.util.Texts
import net.minecraft.server.command.CommandSource
import net.minecraft.util.Formatting

/**
 * Created by 086 on 17/11/2017.
 */
object ToggleCommand : Command() {
    override fun register(dispatcher: CommandDispatcher<CommandSource>) {
        dispatcher.register(
            LiteralArgumentBuilder.literal<CommandSource>("toggle")
                .then(
                    RequiredArgumentBuilder.argument<CommandSource, Module>(
                        "module",
                        ModuleArgumentType.module()
                    )
                        .executes { context: CommandContext<CommandSource> ->
                            val m =
                                context.getArgument(
                                    "module",
                                    Module::class.java
                                )
                            m.toggle()
                            (context.source as KamiCommandSource).sendFeedback(
                                Texts.f(
                                    Formatting.GOLD, Texts.append(
                                        Texts.lit("Toggled module "),
                                        Texts.flit(
                                            Formatting.YELLOW,
                                            m.name.value
                                        ),
                                        Texts.lit(", now "),
                                        Texts.flit(
                                            if (m.isEnabled) Formatting.GREEN else Formatting.RED,
                                            if (m.isEnabled) "ON" else "OFF"
                                        )
                                    )
                                )
                            )
                            0
                        }
                )
        )
    }
}